### PR TITLE
docs: add how to write a userstyle guide

### PR DESCRIPTION
**Pull request by uncenter**
I'm terrible at writing and I'm not sure if this is at all something you guys would be interested in, but I thought it could be nice to have a detailed walkthrough of the template and what it could/should look like.

### CODE REVIEW

File: docs/how-to-write-a-userstyle.md [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+
+### Setup
+
+Assuming you have completed steps 1-6 of [Creating Userstyles](./userstyle-creation.md#creating-userstyles), you should have the template somewhat filled out. In this guide we'll be theming [example.org](http://example.org/), a _very_ important website that I use all the time.
+
+#### Usercss Metadata
+
+I'll start by filling out the "metadata block" at the very top of the template. Replace `<port name>` with the name of the website you are porting, and please ensure it is capitalized when needed. For `example.org`, it looks like this:
+
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:26]:
```suggestion
```less
```

File: docs/how-to-write-a-userstyle.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+==/UserStyle== */
+```
+
+The rest of the metadata block details the preprocessor and the options of the userstyle. You won't need to change any of this (other than `<port name>`), but if you are curious you can [read the specification](https://github.com/openstyles/stylus/wiki/Writing-UserCSS#usercss-metadata).
+
+#### `@-moz-document`
+
+Next, we'll want to tell Stylus to apply this userstyle on our specific website. Just replace `<domain of your website>` with... you guess it! The domain of your website. For our `example.org` example, it looks like this:
+
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:50]:
```suggestion
```less
```

File: docs/how-to-write-a-userstyle.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+```
+
+> [!NOTE]
+> While `domain('<domain>')` works for most cases, you may need to use [`regexp('<your regex>')`](https://github.com/openstyles/stylus/wiki/Writing-styles#regexp).
+
+#### The lookup table
+
+The next section of the template — we'll refer to it as the "lookup table — doesn't need any changes from you,, as it just contains the Catppuccin palette for each of the flavors.
+
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:61]:
```suggestion
```less
```

File: docs/how-to-write-a-userstyle.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    @mocha:     { @rosewater: #f5e0dc; @flamingo: #f2cdcd; @pink: #f5c2e7; @mauve: #cba6f7; @red: #f38ba8; @maroon: #eba0ac; @peach: #fab387; @yellow: #f9e2af; @green: #a6e3a1; @teal: #94e2d5; @sky: #89dceb; @sapphire: #74c7ec; @blue: #89b4fa; @lavender: #b4befe; @text: #cdd6f4; @subtext1: #bac2de; @subtext0: #a6adc8; @overlay2: #9399b2; @overlay1: #7f849c; @overlay0: #6c7086; @surface2: #585b70; @surface1: #45475a; @surface0: #313244; @base: #1e1e2e; @mantle: #181825; @crust: #11111b; };
+  }
+```
+
+### The `#catppuccin` mixin
+
+We'll refer to this next section of the template as the "`#catppuccin` mixin". You can ignore the first half of this section with all of the `@<color>: @catppuccin[@@lookup][@<color>];` lines and skip right to the comment where it says `// Start styling your website here:`. As the comment suggests, you can start adding your styles here. For `example.org`, I added a few basic declarations for colors and background colors:
+
+<!-- prettier-ignore -->
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:76]:
```suggestion
```less
```

File: docs/how-to-write-a-userstyle.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+    /* ... */
+```
+
+#### Applying the theme
+
+The final section of the template is about applying the user's light/dark flavors. Read the comments thoroughly and decide which of the two options works best for your site. In our case, `example.org` doesn't have a light/dark mode toggle, so we'll just apply it based on the user's preferences.
+
+<!-- prettier-ignore -->
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:103]:
```suggestion
```less
```

File: docs/how-to-write-a-userstyle.md [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+      #catppuccin(@darkFlavor, @accentColor);
+    }
+  }
+```
+
+#### Putting it all together
+
+If we combine all of the previous steps, we have a working userstyle!
+
+```css
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:120]:
```suggestion
```less
```

File: docs/images/example.org-original.png
Comment (isabelroses) [inline docs/images/example.org-original.png:1]: This should be a webp

File: docs/images/example.org-themed.png
Comment (isabelroses) [inline docs/images/example.org-themed.png:1]: This should be a webp

**Review by isabelroses — CHANGES_REQUESTED** [review]

File: docs/how-to-write-a-userstyle.md
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:1]: `<port name>`, `<domain of your website>` and so on, in my opinon, is worded a bit strangly, I think simply putting "your port" or something works just as well.

File: docs/how-to-write-a-userstyle.md [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+    @macchiato: { @rosewater: #f4dbd6; @flamingo: #f0c6c6; @pink: #f5bde6; @mauve: #c6a0f6; @red: #ed8796; @maroon: #ee99a0; @peach: #f5a97f; @yellow: #eed49f; @green: #a6da95; @teal: #8bd5ca; @sky: #91d7e3; @sapphire: #7dc4e4; @blue: #8aadf4; @lavender: #b7bdf8; @text: #cad3f5; @subtext1: #b8c0e0; @subtext0: #a5adcb; @overlay2: #939ab7; @overlay1: #8087a2; @overlay0: #6e738d; @surface2: #5b6078; @surface1: #494d64; @surface0: #363a4f; @base: #24273a; @mantle: #1e2030; @crust: #181926; };
+    @mocha:     { @rosewater: #f5e0dc; @flamingo: #f2cdcd; @pink: #f5c2e7; @mauve: #cba6f7; @red: #f38ba8; @maroon: #eba0ac; @peach: #fab387; @yellow: #f9e2af; @green: #a6e3a1; @teal: #94e2d5; @sky: #89dceb; @sapphire: #74c7ec; @blue: #89b4fa; @lavender: #b4befe; @text: #cdd6f4; @subtext1: #bac2de; @subtext0: #a6adc8; @overlay2: #9399b2; @overlay1: #7f849c; @overlay0: #6c7086; @surface2: #585b70; @surface1: #45475a; @surface0: #313244; @base: #1e1e2e; @mantle: #181825; @crust: #11111b; };
+  }
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:68]:
```suggestion
}
```

File: docs/how-to-write-a-userstyle.md [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+        color: @text;
+        background-color: @base;
+      }
+    }
+
+    a:link, a:visited {
+      color: @blue;
+    }
+
+    /* ... */
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:95]:
```suggestion
#catppuccin(@lookup, @accent) {
  @rosewater: @catppuccin[@@lookup][@rosewater];
  /* ... */
  @accent-color: @catppuccin[@@lookup][@@accent];

  body {
    background-color: @surface0;

    & > div {
      color: @text;
      background-color: @base;
    }
  }

  a:link, a:visited {
    color: @blue;
  }

  /* ... */
```

File: docs/how-to-write-a-userstyle.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+  @media (prefers-color-scheme: light) {
+    :root {
+      #catppuccin(@lightFlavor, @accentColor);
+    }
+  }
+  @media (prefers-color-scheme: dark) {
+    :root {
+      #catppuccin(@darkFlavor, @accentColor);
+    }
+  }
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:113]:
```suggestion
@media (prefers-color-scheme: light) {
  :root {
    #catppuccin(@lightFlavor, @accentColor);
  }
}
@media (prefers-color-scheme: dark) {
  :root {
    #catppuccin(@darkFlavor, @accentColor);
  }
}
```

File: docs/how-to-write-a-userstyle.md
Comment (uncenter) [reply docs/how-to-write-a-userstyle.md:1]: I'm only using `<port name>` where it refers to literally replacing that string in the template though, and same with the other example.

File: template/catppuccin.user.css [hunk 10]
```diff
@@ -77,29 +70,29 @@
...
     }
   }
+  */
 
-  // manually set colorscheme
-  .light-theme {
+  /* If the website you are theming *does* have a light/dark mode toggle,
+     you can conditionally enable light/dark flavors by targeting whatever attribute that website uses to indicate the theme.
+     Un-comment this block if that is the case.
+  :root[data-theme="dark"] {
```
Comment (isabelroses) [inline template/catppuccin.user.css:90]: I think leaving a reference to https://github.com/catppuccin/userstyles/blob/39288b593278e5efa48007d5fc001292cf910672/styles/youtube/catppuccin.user.css#L21-L30

And perhaps a reference to https://github.com/catppuccin/userstyles/blob/39288b593278e5efa48007d5fc001292cf910672/styles/brave-search/catppuccin.user.css#L16-L29

Or these could be placed in the `how-to-write-a-userstyle.md` file.

File: docs/how-to-write-a-userstyle.md
Comment (isabelroses) [reply docs/how-to-write-a-userstyle.md:1]: I mean it makes sense, but it feels like a over complex/obstructive way to refer to it, but that's just my opinion.

**Review by isabelroses — CHANGES_REQUESTED** [review]

File: template/catppuccin.user.css (hunk 10 above)
Comment (uncenter) [reply template/catppuccin.user.css:90]: Ohh yeah smart, I'll do that.

File: docs/how-to-write-a-userstyle.md
Comment (uncenter) [reply docs/how-to-write-a-userstyle.md:1]: It doesn't seem very complex to me, rather just very clear as to what parts to change.

> Replace `<port name>` with the name of the website you are porting

If I said "Replace your port with the name of the website you are porting" that makes no sense.

**Comment by sgoudham** [comment]
I quite like GitHub's new [Writing Content To Be Translated](https://docs.github.com/en/contributing/writing-for-github-docs/writing-content-to-be-translated) page that covers some basic fundamentals/priciples for technical writing, we should take that into consideration here.

**Comment by uncenter** [comment]
> I quite like GitHub's new [Writing Content To Be Translated](https://docs.github.com/en/contributing/writing-for-github-docs/writing-content-to-be-translated) page that covers some basic fundamentals/priciples for technical writing, we should take that into consideration here.
> 
> See also: [Microsoft's Writing Style Guide](https://learn.microsoft.com/en-us/style-guide/bias-free-communication)

Thank you, I know my writing is not great... appreciate the links. I don't mean we need to implement this exact suggestion but I'm just trying to prompt us somewhere in regards to documentation (which currently feels pretty lackluster).

File: docs/how-to-write-a-userstyle.md
Comment (isabelroses) [reply docs/how-to-write-a-userstyle.md:1]: I see what you mean

File: docs/how-to-write-a-userstyle.md [hunk 11]
```diff
@@ -0,0 +1,208 @@
...
+    @mocha:     { @rosewater: #f5e0dc; @flamingo: #f2cdcd; @pink: #f5c2e7; @mauve: #cba6f7; @red: #f38ba8; @maroon: #eba0ac; @peach: #fab387; @yellow: #f9e2af; @green: #a6e3a1; @teal: #94e2d5; @sky: #89dceb; @sapphire: #74c7ec; @blue: #89b4fa; @lavender: #b4befe; @text: #cdd6f4; @subtext1: #bac2de; @subtext0: #a6adc8; @overlay2: #9399b2; @overlay1: #7f849c; @overlay0: #6c7086; @surface2: #585b70; @surface1: #45475a; @surface0: #313244; @base: #1e1e2e; @mantle: #181825; @crust: #11111b; };
+}
+```
+
+### The `#catppuccin` mixin
+
+We'll refer to this next section of the template as the "`#catppuccin` mixin". You can ignore the first half of this section with all of the `@<color>: @catppuccin[@@lookup][@<color>];` lines and skip to the comment where it says `// Start styling your website here:`. As the comment suggests, you will write the CSS rules for your port here. Here is what it looks like for our `example.org` port:
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:73]:
```suggestion
We'll refer to this next section of the template as the "`#catppuccin` mixin". You can ignore the first half of this section with all the `@<color>: @catppuccin[@@lookup][@<color>];` lines and skip to the comment where it says `// Start styling your website here:`. As the comment suggests, you will write the CSS rules for your port here. Here is what it looks like for our `example.org` port:
```

File: docs/how-to-write-a-userstyle.md [hunk 12]
```diff
@@ -0,0 +1,208 @@
...
+@media (prefers-color-scheme: dark) {
+  :root {
+    #catppuccin(@darkFlavor, @accentColor);
+  }
+}
+```
+
+#### Putting it all together
+
+Combining all of the previous steps, we have a working userstyle!
```
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:118]:
```suggestion
Combining all the previous steps, we have a working userstyle!
```

**Review by isabelroses — CHANGES_REQUESTED** [review]

File: docs/how-to-write-a-userstyle.md (hunk 12 above)
Comment (uncenter) [reply docs/how-to-write-a-userstyle.md:118]: I think this is just a preference with no correct answer. I'm more of a fan of the more explicit version with "of" as you can see but...

File: docs/how-to-write-a-userstyle.md (hunk 11 above)
Comment (uncenter) [reply docs/how-to-write-a-userstyle.md:73]: Same as https://github.com/catppuccin/userstyles/pull/353#discussion_r1423380271.

File: docs/how-to-write-a-userstyle.md [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+### Prerequisites
+
+- [Stylus](https://github.com/openstyles/stylus) ([Firefox](https://addons.mozilla.org/en-GB/firefox/addon/styl-us/) / [Chrome](https://chrome.google.com/webstore/detail/stylus/clngdbkpkpeebahjckkjfobafhncgmne)
+  extension).
+- An IDE (e.g. [Visual Studio Code](https://code.visualstudio.com/))
+- A website to theme!
+
+### Setup
+
+Assuming you have completed steps 1-6 of [Creating Userstyles](./userstyle-creation.md#creating-userstyles), you should have the template somewhat filled out. Follow along with your userstyle as we write one for [example.org](http://example.org/).
```
Comment (uncenter) [inline docs/how-to-write-a-userstyle.md:20]: Hmm I probably don't need this since we walk through this part in the next step...

Comment (uncenter) [reply docs/how-to-write-a-userstyle.md:20]: Well steps 1-4 *are* relevant. Maybe I could literally go through each and every step on that list (https://github.com/catppuccin/userstyles/blob/main/docs/userstyle-creation.md#creating-userstyles)?

That would mean adding sections on adding a new userstyle to `userstyles.yml` and creating image previews (this is an important one - we should walk through taking good screenshots, converting to WEBP, and of course using catwalk).

Comment (isabelroses) [reply docs/how-to-write-a-userstyle.md:20]: I thinks its alright, just remove "steps 1-6 of"

**Comment by isabelroses** [comment]
PR will need refactoring as of https://github.com/catppuccin/userstyles/pull/400 being merged

File: docs/how-to-write-a-userstyle.md
Comment (isabelroses) [inline docs/how-to-write-a-userstyle.md:1]: This file does not contain `@supportURL`

**Review by isabelroses — CHANGES_REQUESTED** [review]

**Comment by thelooter** [comment]
Overall I think this guide is really helpful. I'd also suggest adding a mention of needing to enable usercss while creating the style in stylus. I searched for that checkbox for almost 30 minutes. It#s not super apparent

**Comment by uncenter** [comment]
> Overall I think this guide is really helpful. I'd also suggest adding a mention of needing to enable usercss while creating the style in stylus. I searched for that checkbox for almost 30 minutes. It#s not super apparent

Screenshot 2023-12-17 at 11 28 58 (Arc)
This checkbox right?

**Comment by thelooter** [comment]
Exactly. That's the checkbox I meant

**Comment by uncenter** [comment]
> Exactly. That's the checkbox I meant

Hmm where in the guide should I mention it?

File: docs/images/write-new-style-with-usercss.png
Comment (isabelroses) [inline docs/images/write-new-style-with-usercss.png:1]: can you use a webp instead.

Comment (uncenter) [reply docs/images/write-new-style-with-usercss.png:1]: Oop yeah I forgot. It was just a temporary thing since I haven't even added it in any docs files.

**Comment by uncenter** [comment]
Remaining things...
- [ ] Add details on how to set up live reloading with Stylus, how to create a new userstyle with UserCSS on Stylus.
- [ ] Add guide on using Catwalk and taking good looking screenshots (i.e. Firefox's "Save Visible" mode for screenshots to keep everything consistent).

**Comment by isabelroses** [comment]
> * [ ]  Add guide on using Catwalk and taking good looking screenshots (i.e. Firefox's "Save Visible" mode for screenshots to keep everything consistent).

I believe there's no need to write the Catwalk docs two times, just direct the users to Catwalk, but be my guest to educate on good screenshots and so on.

**Comment by uncenter** [comment]
I found the Catwalk documentation extremely insufficient, at least when I first started. Looks better now, still confusing though... doesn't mention at all converting to WebP, which is an important step that Catwalk doesn't mention.

**Comment by isabelroses** [comment]
> I found the Catwalk documentation extremely insufficient, at least when I first started. Looks better now, still confusing though... doesn't mention at all converting to WebP, which is an important step that Catwalk doesn't mention.

Then I would argue that's better for upstream rather then here.

**Comment by uncenter** [comment]
> > I found the Catwalk documentation extremely insufficient, at least when I first started. Looks better now, still confusing though... doesn't mention at all converting to WebP, which is an important step that Catwalk doesn't mention.
> 
> Then I would argue that's better for upstream rather then here.

Then I would argue you are completely right and I will go do that. 🫡

**Comment by GitMuslim** [comment]
image

@uncenter https://github.com/gitmuslim/example

**Review by sgoudham — APPROVED** [review]

**Review by isabelroses — APPROVED** [review]

**Review by uncenter — COMMENTED** [review]
LGTM.

**Pull request merged**